packages/cli/index.js

```diff
@@ -172,27 +194,31 @@ program
           return site.rebuildAffectedSourceFiles(filePath);
         }
         return site.buildAsset(filePath);
-      }).catch((err) => {
-        logger.error(err.message);
-      });
+      })
+        .then(() => onePagePath && options.backgroundBuild && backgroundBuildAndReload())
+        .catch((err) => {
+          logger.error(err.message);
+        });
     };
 
     const removeHandler = (filePath) => {
       logger.info(`[${new Date().toLocaleTimeString()}] Reload for file deletion: ${filePath}`);
-      logger.info('Synchronizing opened pages list before reload');
-      const normalizedActiveUrls = liveServer.getActiveUrls().map((url) => {
-        const completeUrl = path.extname(url) === '' ? path.join(url, 'index') : url;
-        return fsUtil.removeExtension(completeUrl);
-      });
-      site.changeCurrentOpenedPages(normalizedActiveUrls);
+      if (onePagePath) {
+        if (options.backgroundBuild) {
+          site.stopOngoingBuilds();
+        }
+        syncOpenedPages();
+      }
       Promise.resolve('').then(() => {
         if (site.isFilepathAPage(filePath) || site.isDependencyOfPage(filePath)) {
           return site.rebuildSourceFiles(filePath);
         }
         return site.removeAsset(filePath);
-      }).catch((err) => {
-        logger.error(err.message);
-      });
+      })
+        .then(() => onePageWithBackgroundBuild && backgroundBuildAndReload())
```

Comment:
From a concerns standpoint (minimising site building organisation in `cli` package), could we move these into `Site`'s methods? (`rebuildSourceFiles / regenerateAffectedPages`)

Might be a little awkward with the `liveServer.reloadActiveTabs();` call though, also not a huge leakage - I'll leave it up to you!

Reply:
Yeah I feel the same as you, but `backgroundBuildNotViewedFiles` is run not just on `rebuildSourceFiles / regenerateAffectedPages`, but also on `buildAsset / removeAsset` as well. Putting the background build call on the asset methods is a bit awkward, so I decided to put it outside the methods on the `cli` package, though not ideal.

Plus, it can use `live-server`s methods without porting over to `Site` (otherwise, we need to add it as a dependency and have a copy of the patch there, which is redundant, and I'm not really confident in making `Site` share concerns of live-server's capabilities).

Reply:
> but also on buildAsset / removeAsset as well

👀 ahh good catch, could we tighten this actually? since assets shouldn't affect page generation, shouldn't need to stop - restart ongoing builds.

> Plus, it can use live-servers methods without porting over to Site (otherwise, we need to add it as a dependency and have a copy of the patch there, which is redundant, and I'm not really confident in making Site share concerns of live-server's capabilities).

Was thinking of just passing a callback; might be a little awkward though per your second point.

Reply:
I see, then as assets build shouldn't affect generation, I can move the background build call to the respective functions.

However, as those are `delay`-ed functions, they are expected to receive only a single argument which is an array (if needed). Then putting the callback to promise is not an option per the spec. Maybe I can keep the `liveServer.reloadActiveTabs()` call in `cli` but after `await`-ed the call to the two functions, do you think that would be better?

Reply:
> Maybe I can keep the liveServer.reloadActiveTabs() call in cli (guarded by a conditional check to -b) but only invoke it after await-ed the call to the corresponding functions, do you think that would be better?

Yes, this looks good as well!

I was thinking passing the callback as a constructor argument to `Site` (injecting `reloadAfterBackgroundBuild` as a black box dependency to `Site`) earlier.

The issue might be that we have been handling `live-server` specifics from `cli` package in the first place for a long time now. So there's quite a bit of communication going on between the two packages.
Perhaps we could consider moving the entire `live-server` specifics to the `core` package (or another package? e.g. `@markbind/dev-server`) **much later**?

Would that be better?

Reply:
> Yes, this looks good as well!
>
> I was thinking passing the callback as a constructor argument to Site (injecting reloadAfterBackgroundBuild as a black box dependency to Site) earlier.

edit: realised the latter might be a little more suitable for now, as there's a couple of places where `backgroundBuildNotViewedFiles` is / should be called (`reloadSiteConfig` / `generatePages` / `rebuiltAffected` / `rebuildSource`)
-- Is it possible to have a single call to the `reloadAfterBackgroundBuild` inside `backgroundBuildNotViewedFiles`? (and then call `backgroundBuildNotViewedFiles` at these places, instead of repeating `reloadAfterBackgroundBuild` in `cli` package multiple times)

Reply:
I see. I guess for now we can pass it as a constructor argument and have the reload all tabs happen in `Site`'s methods.

> Perhaps we could consider moving the entire live-server specifics to the core package (or another package? e.g. @markbind/dev-server) much later?

Moving forward I think we can move the server operations to a package as suggested for better separation of concerns.